File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.js
const BFS = (row, col, array) => {
    // 상하좌우
    const dy = [1, -1, 0, 0];
    const dx = [0, 0, 1, -1];
    
    const visited = Array.from({length: 5}, () => new Array(5).fill(false));
    const queue = [[row, col, 0]];
    visited[row][col] = true;
    
    while (queue.length) {
        const [y, x, dist] = queue.shift();
        
        if (dist > 1) continue; // 거리 2 이상은 탐색 X
        
        // 탐색 (상하좌우)
        for (let i = 0; i < 4; i++) {
            const ny = y + dy[i];
            const nx = x + dx[i];
            
            // 범위 확인
            if (ny >= 0 && ny < 5 && nx >= 0 && nx < 5 && !visited[ny][nx]) {
                if (array[ny][nx] === "P") {    // 응시자 인접
                    return 0;
                } else if (array[ny][nx] === "O") { // 빈 테이블 인접
                    // 빈 테이블의 상하좌우에 P 존재하면 안됨
                    queue.push([ny, nx, dist + 1]);
                }
            }
        }
    }
    
    return 1;
}

function solution(places) {
    const answer = [];
    
    places.forEach(place => {
        place = place.map(row => row.split(''));
        let isSafe = true;
        
        // BFS
        for (let i = 0; i < 5; i++) {
            for (let j = 0; j < 5; j++) {
                if (place[i][j] === "P") {
                    if (BFS(i, j, place) === 0) {
                        isSafe = false;
                        break;
                    }
                }
            }
            if (!isSafe) break;
        }
        
        isSafe ? answer.push(1) : answer.push(0);
    })
    
    return answer;
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

def BFS(row, col, array):
    # 상하좌우
    dy = [1, -1, 0, 0]
    dx = [0, 0, 1, -1]
    
    visited = [[False] * 5 for _ in range(5)]
    queue = deque([[row, col, 0]])    # 행, 열, 거리
    visited[row][col] = True;
    
    while queue:
        y, x, dist = queue.popleft()
        
        # 거리 2 이상은 탐색 X
        if dist > 1:
            continue
            
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            
            # 범위, 방문여부 체크
            if 0 <= ny <= 4 and 0 <= nx <= 4 and not visited[ny][nx]:
                visited[ny][nx] = True; # 방문처리
                
                if array[ny][nx] == "P":
                    return 0
                elif array[ny][nx] == "O":
                    queue.append([ny, nx, dist + 1])
            
    return 1    # 인접 요소가 파티션일 때 포함

def solution(places):
    answer = []
    
    for place in places:
        place = [list(row) for row in place]
        is_safe = True
        
        # P인 요소만 탐색
        for i in range(5):
            for j in range(5):
                if place[i][j] == "P":
                    if BFS(i, j, place) == 0:
                        is_safe = False
                        break   # 내부 반복문 종료
            if not is_safe:
                break   # 외부 반복문 종료
        
        answer.append(1 if is_safe else 0)
        
    return answer